app/src/main/java/io/homeassistant/companion/android/settings/sensor/views/SensorUpdateFrequencyView.kt

```diff
@@ -0,0 +1,76 @@
+package io.homeassistant.companion.android.settings.sensor.views
+
+import android.os.Build
+import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.layout.Row
+import androidx.compose.foundation.layout.padding
+import androidx.compose.foundation.rememberScrollState
+import androidx.compose.foundation.verticalScroll
+import androidx.compose.material.Divider
+import androidx.compose.material.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.platform.LocalContext
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.unit.dp
+import io.homeassistant.companion.android.common.R
+import io.homeassistant.companion.android.common.sensors.BatterySensorManager
+import io.homeassistant.companion.android.common.sensors.SensorWorkerBase
+import io.homeassistant.companion.android.database.AppDatabase
+import io.homeassistant.companion.android.database.settings.SensorUpdateFrequencySetting
+import io.homeassistant.companion.android.util.compose.InfoNotification
+import io.homeassistant.companion.android.util.compose.RadioButtonRow
+
+@Composable
+fun SensorUpdateFrequencyView(
+    sensorUpdateFrequency: SensorUpdateFrequencySetting,
+    onSettingChanged: (SensorUpdateFrequencySetting) -> Unit
+) {
+    val scrollState = rememberScrollState()
+    val context = LocalContext.current
+    Column(
+        modifier = Modifier
+            .padding(20.dp)
+            .verticalScroll(scrollState)
+    ) {
+        Row(
+            verticalAlignment = Alignment.CenterVertically,
+            modifier = Modifier.padding(bottom = 20.dp)
+        ) {
+            Text(stringResource(R.string.sensor_update_frequency_description))
+        }
+        Divider()
+        RadioButtonRow(
+            text = stringResource(R.string.sensor_update_frequency_normal),
+            selected = sensorUpdateFrequency == SensorUpdateFrequencySetting.NORMAL,
+            onClick = { onSettingChanged(SensorUpdateFrequencySetting.NORMAL) }
+        )
+        val isCharging = AppDatabase.getInstance(context).sensorDao().get(BatterySensorManager.isChargingState.id)
+        if (isCharging != null && isCharging.enabled)
```

Comment:
I don't think hiding the setting if the sensor has been turned off is a good idea, because
- it will make it harder to discover that this is an option, which might be a reason for users to turn on the required sensor
- if the user previously selected fast while charging, and then disabled the sensor, the UI will show the setting as fast always while it hasn't actually changed and will change back once the sensor is re-enabled

Personally, I would show a warning (different text?) or disable the option (greyed out) instead of hiding it when the required sensor is not activated. Falling back to normal update frequency instead of always fast updating also makes more sense to me, as it is the 'safer'/default option and users would have selected always if they wanted that.

Reply:
> I don't think hiding the setting if the sensor has been turned off is a good idea, because
> 
> * it will make it harder to discover that this is an option, which might be a reason for users to turn on the required sensor
> * if the user previously selected fast while charging, and then disabled the sensor, the UI will show the setting as fast always while it hasn't actually changed and will change back once the sensor is re-enabled
> 
> Personally, I would show a warning (different text?) or disable the option (greyed out) instead of hiding it when the required sensor is not activated. 

I like disabling the option, I think that makes more sense instead of hiding it.

> Falling back to normal update frequency instead of always fast updating also makes more sense to me, as it is the 'safer'/default option and users would have selected always if they wanted that.

So the reason why it falls back to Fast Always is because once a user selects the fast option we automatically register the intent for `TIME_TICK` on the next app start. I didn't want to change the setting when that happens because that means we need to add logic to the sensor detail screen to change the DB value but also to warn them that they need to restart the application to unregister the intent.  When a user enables this option (Fast While Charging) they already see the text saying the sensor is required to be enabled so I thought that was enough of a warning there and at this point they already had it enabled because they see the option. So disabling the sensor becomes a conscious choice. I would like to avoid having to overcomplicate the logic on the sensor detail screen especially given that the current logic needs to be switched to Compose eventually.

Do you think stronger wording and disabling instead of hiding will make it more clear?

I suppose we could also change the text for Normal to say "most battery friendly" as well?

Reply:
> So the reason why it falls back to Fast Always (...)

I understand why it's happening from a technical point of view, but from the user's point of view unknown charging status != charging (otherwise the option should be called fast except when unplugged).

> When a user enables this option (Fast While Charging) they already see the text (...) So disabling the sensor becomes a conscious choice.

Maybe when disabling it right now, but if they disable it a few months later? This is the only non-sensor setting in the app to use a sensor, and the update frequency could also have [read the API directly](https://developer.android.com/reference/android/os/BatteryManager#isCharging()) on most devices so this might be unexpected if they don't remember all the text for this setting (see: the websocket service doesn't require you to enable the interactive sensor).

> Do you think stronger wording and disabling instead of hiding will make it more clear?

I think the confusion can be solved if the 'while charging' setting text is changed when the sensor is disabled, by removing the last sentence and placing more info on a new line which will make it stand out:
- If not selected: turn on the sensor to use this option
- If selected: the required sensor is disabled, currently always updating fast

> I suppose we could also change the text for Normal to say "most battery friendly" as well?

Updating less / only while charging / ... is even more battery friendly, I think 'Normal' is fine.

Reply:
> Maybe when disabling it right now, but if they disable it a few months later? This is the only non-sensor setting in the app to use a sensor, and the update frequency could also have [read the API directly](https://developer.android.com/reference/android/os/BatteryManager#isCharging()) on most devices so this might be unexpected if they don't remember all the text for this setting (see: the websocket service doesn't require you to enable the interactive sensor).

if only that API was available in API 21



> I think the confusion can be solved if the 'while charging' setting text is changed when the sensor is disabled, by removing the last sentence and placing more info on a new line which will make it stand out:
> 
> * If not selected: turn on the sensor to use this option
> * If selected: the required sensor is disabled, currently always updating fast

Good call, will update the option text to better reflect this. Personally I don't think too many people will be disabling this sensor as its enabled by default and a pretty common use case.

Reply:
> if only that API was available in API 21

I know and that is why I used "on most devices". However it turns out the intent is sticky so there is actually an option for this that should work on API 21, I updated my comment a few minutes later but apparently too late :)

Reply:
> > if only that API was available in API 21
> 
> I know and that is why I used "on most devices". However it turns out the intent is sticky so there is actually an option for this that should work on API 21, I updated my comment a few minutes later but apparently too late :)

funny thats exactly what our BatterySensorManager does today, as its a sticky intent thats much better to go by I think. Let me run some tests...this will simplify things even more. Good call :)

Reply:
ok removed that requirement and updated the text, thanks for the suggestion :)